### PR TITLE
Also consider requestedPermissions when runtime permission not granted (on Kitkat+ devices)

### PR DESCRIPTION
**Pull request by marcardar**
My hunch about inspecting the package manager for the requested permissions turned out to be correct.

Then I saw someone else actually [reported this](https://code.google.com/p/android/issues/detail?id=25906#c16) in the Android Issue tracker.  More details [here](http://stackoverflow.com/questions/18573139/programmatically-alter-manifest-android-custom-permissions).  Basically, the only issue is that someone may have rooted and then denied the client app permission intentionally. Surely we can ignore this?

### CODE REVIEW

**Comment by timrae** [comment]
I'm not sure this is a good idea tbh. Merging this change would lead to a flow like this:

1) User installs client app. READ_WRITE_DATABASE has not been defined in the system yet, so the user is not warned about the AnkiDroid permission

2) User installs AnkiDroid, puts some effort in making their cards

3) User opens client app, and does something to activate the API. 

4) The provider checks the manifest of the client app and sees that it has the permission declared (though user doesn't know anything about this) and grants full access

5) Client app replaces all the users fields with empty strings

6) User is outraged that we gave the client app edit access to their collection without their knowledge

**Comment by marcardar** [comment]
For installs from Google Play, the permissions are shown from the Play Store app (when the install is triggered), not when the app is actually installed.  So it shows the dangerous permission regardless of whether AnkiDroid is already installed or not.

**Comment by marcardar** [comment]
So the problem is for non Google Play installs (I checked that direct installs do not show permission when AnkiDroid not installed). How about we compare the install times?

**Comment by timrae** [comment]
Nah, Google Play doesn't show custom permissions. [Aedict3](https://play.google.com/store/apps/details?id=sk.baka.aedict3&hl=en) for example definitely declares our permission and it doesn't show up there under permissions.

**Comment by marcardar** [comment]
Oh that's very interesting!  Google Play shows the custom permission only if AnkiDroid is already installed. I never knew that.

**Comment by marcardar** [comment]
device-2016-03-17-223956

**Comment by timrae** [comment]
The AnkiDroid manifest has the title and description for the custom permission so it'd be pretty hard for them to do it without AnkiDroid being installed unless they kept a database of custom permissions on their website which would probably not be all that reliable.

**Comment by marcardar** [comment]
Yeah, it makes sense.

So the problem is that when the app doesn't have permission, then the user has not seen the permission, therefore we need to get the user to see the permission.

**Comment by marcardar** [comment]
How about an Activity that asks the user whether it's okay to grant permissions to a certain app?

**Comment by marcardar** [comment]
This is going to be a major problem, because client will ask user to install AnkiDroid and then AnkiDroid import will not work because it doesn't have permission.

**Comment by marcardar** [comment]
> 5) Client app replaces all the users fields with empty strings
> 6) User is outraged that we gave the client app edit access to their collection without their knowledge

If this happens, then it will just as easily happen to apps that the user has already granted access to.  The client app replacing fields is 99% a problem with the client app.

**Comment by timrae** [comment]
I think it's OK how it is. Everything except for note update and deletion can be done without any permission on Lollipop and below. Updates and deletes will work if the user installs AnkiDroid first, or after the first update of the client app.

**Comment by marcardar** [comment]
Yeah, it's not so bad that way.  For the updates, if the user is prompted to reinstall they often get nervous they will lose their data (e.g. starred words). Is it possible that when the client app is updated, it gets the permission? Have you tested that?

**Comment by marcardar** [comment]
Regardless, for apps that neither have the permission granted nor requested, we should deny them query/insert too.

**Comment by marcardar** [comment]
I'll update this PR accordingly.

**Comment by marcardar** [comment]
Done - please check.

**Comment by marcardar** [comment]
I think the update() method needs to be, er, updated too.

**Comment by marcardar** [comment]
Done

**Comment by marcardar** [comment]
Tidied up the code (removed a redundant condition check) by making the code not differ so much from current version.

**Comment by marcardar** [comment]
And now also allows testing code to pass on query/insert!

**Comment by marcardar** [comment]
I've refactored to make it all clearer.

I also tested update app-updates - yes, when the user updates the client app and the permission was not previously granted (but AnkiDroid app has since been installed), then the installation will show the AnkiDroid permission.  Not sure this helps much (saying to the users, hang back until next update!) but it's good to know anyway.

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 1]
```diff
@@ -1181,7 +1180,30 @@ private void logProviderCall(String methodName, Uri uri) {
         }
     }
 
-    private boolean hasReadWritePermission() {
+    /**
+     * Provide minimal protection for query/insert because these cannot result in the loss of user data
+     */
+    private boolean allowQueryInsert() {
```
Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: Could change this to "allowNonDestructiveChanges()" (and "allowDestructiveChanges()")

**Comment by timrae** [comment]
I structured my previous PR the way I did on purpose as I want to have the control logic inline in the main methods. I feel this PR makes it less clear what's happening when reading through the methods.

I would strongly prefer only having one new method, e.g. by changing `hasRequestedReadWritePermission()` to something like this:

```java
    /** Returns true if the calling package is known to be 'rogue' and should be blocked.
         A calling package might be rogue if it hasn't declared #READ_WRITE_PERMISSION in 
         its manifest, or if explicitly blacklisted. **/
    private boolean knownRogueClient() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            return false;    // no way to ascertain whether rogue client app
        }
        PackageManager pm = getContext().getPackageManager();
        try {
            PackageInfo callingPi = pm.getPackageInfo(getCallingPackage(), PackageManager.GET_PERMISSIONS);
            return callingPi != null && callingPi.requestedPermissions != null
                    && Arrays.asList(callingPi.requestedPermissions).contains(READ_WRITE_PERMISSION);
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 2]
```diff
@@ -1181,7 +1180,30 @@ private void logProviderCall(String methodName, Uri uri) {
...
+    }
+
+    /**
+     * Provide more protection for update/delete because these can result in the loss of user data
+     */
+    private boolean allowUpdateDelete() {
+        return hasReadWritePermissionOrCalledByUnitTests();
+    }
+
+    private boolean hasReadWritePermissionOrCalledByUnitTests() {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: This isn't necessary. As I have previously said, `getContext().getPackageName().equals(getCallingPackage())` implies READ_WRITE_PERMISSION is granted.

**Comment by timrae** [comment]
Fixed bug in above snipped

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java (hunk 2 above)
Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: If this is true, then we don't need the Build.DEBUG check.

**Comment by marcardar** [comment]
> private boolean knownRogueClient() {

Okay, so this is simply a rename of hasRequestedReadWritePermission(). I guess the debate is whether a method should be named by its function or its intent.  If you name it by its function then you don't need to read its implementation.  The other advantage of hasRequestedReadWritePermission() is it lines up well alongside hasReadWritePermission() (a method which is named according to it's function). The allowX() methods are named according to their intent.

I think the main methods are much clearer in this PR. It genuinely takes more time to parse the old way. The main purpose of the main methods is to to do work:

`if (!hasReadWritePermission() && (CompatHelper.isMarshmallow() || !ok.contains(match)) || knownRogueClient()) {`

`if (ok.contains(match) ? !allowQueryInsert() : !allowUpdateDelete()) {`

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java (hunk 2 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: I'm willing to remove the Build.DEBUG check if you can convincingly prove that the expression

`getContext().checkCallingPermission(READ_WRITE_PERMISSION) == PackageManager.PERMISSION_GRANTED || getContext().getPackageName().equals(getCallingPackage())` returns a different result from `getContext().checkCallingOrSelfPermission(READ_WRITE_PERMISSION) == PackageManager.PERMISSION_GRANTED || getContext().getPackageName().equals(getCallingPackage())` and therefore the warning in the docs about leaking permissions doesn't apply.

Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: Just so I understand:

> As I have previously said, getContext().getPackageName().equals(getCallingPackage()) implies READ_WRITE_PERMISSION is granted.

Is this true regardless of BuildConfig.DEBUG or not?

**Comment by timrae** [comment]
> so this is simply a rename of hasRequestedReadWritePermission()

well it's also a removal of the SDK targeting

> the debate is whether a method should be named by its function

Both your method and the method I suggested is named by its function IMO, and I want to remove the extra methods you introduced that are named by intent

> you don't need to read its implementation

I don't really agree, it's not at all obvious what 'requested' means in `hasRequestedReadWritePermission()`, so developers would still have to read the implementation.

**Comment by timrae** [comment]
`if (ok.contains(match) ? !allowQueryInsert() : !allowUpdateDelete()) {`

This is completely unreadable to me...

**Comment by marcardar** [comment]
Is this more easily readable:

`if (!hasReadWritePermission() && (CompatHelper.isMarshmallow() || !ok.contains(match)) || knownRogueClient()) {`

?

**Comment by marcardar** [comment]
How about this?

`if (ok.contains(match) ? disallowQueryInsert() : disallowUpdateDelete()) {`

?

**Comment by marcardar** [comment]
Or this:

`if (ok.contains(match) ? disallowNonDestructiveChanges() : disallowDestructiveChanges()) {`

?

**Comment by timrae** [comment]
> How about this?

Yeah, I agree it does take a while to parse, but it's quite readable IMO and doesn't require recursing into another method.

> How about this? Or this:

Sorry they're completely unreadable for me. It reads like "If whitelisted then disallow something, otherwise disallow something else"... My brain can't process it.

**Comment by marcardar** [comment]
> Yeah, I agree it does take a while to parse, but it's quite readable IMO and doesn't require recursing into another method.

Three methods...

**Comment by marcardar** [comment]
Just so I understand where you are coming from: you find `!A && (B || !C || D)` "quite readable" but `A ? B : C` "totally unreadable" because of the method names?

**Comment by timrae** [comment]
3 - 4 simple methods like this would also be acceptable for me:

`if (noQueryPermission()) {`
`if (noUpdatePermission()) {`
`if (noDeletePermission()) {`
`if (noInsertPermission()) {`

**Comment by timrae** [comment]
or `if (!allowQuery()) {`... something along those lines

**Comment by marcardar** [comment]
> or if (!allowQuery()) {... something along those lines

This is almost exactly what I had in my original PR :)

**Comment by marcardar** [comment]
Made some changes.

**Comment by timrae** [comment]
This feels like way too much code now.

**Comment by marcardar** [comment]
You can save quite  bit by just moving the "destuctive" named methods into the main methods.

**Comment by timrae** [comment]
Yeah, if you could just change it to [this](https://github.com/ankidroid/Anki-Android/pull/4150#issuecomment-198644468) so we could move onto more important issue it would be appreciated.

**Comment by marcardar** [comment]
Hmm - I've implemented my suggested change. Please take a look.

**Comment by timrae** [comment]
....should I just make my own PR?

**Comment by marcardar** [comment]
I ended up settling on this (example from update() method):

```
private static final List<Integer> NON_DESCTRUCTIVE_UPDATES
        = Arrays.asList(NOTES_ID_CARDS_ORD, MODELS_ID, MODELS_ID_TEMPLATES_ID, SCHEDULE, DECK_SELECTED);
...
if (NON_DESCTRUCTIVE_UPDATES.contains(sUriMatcher.match(uri))) {
    checkNonDestructiveAllowedOrThrow(uri);
}
else {
    checkDestructiveAllowedOrThrow(uri);
}
```

compared to using the existing style:

```
List<Integer> ok = Arrays.asList(NOTES_ID_CARDS_ORD, MODELS_ID, MODELS_ID_TEMPLATES_ID, SCHEDULE, DECK_SELECTED);
if (!hasReadWritePermission() && (CompatHelper.isMarshmallow() || !ok.contains(match)) || knownRogueClient()) {
    throw new SecurityException("Update permission not granted for: " + uri);
}
```

> ....should I just make my own PR?

You're very welcome to!

**Comment by timrae** [comment]
See #4158 which uses much less code

**Pull request closed without merging**